Native compression for all embedded containers

isopov opened the pull request:
Currently to enable compression in the container in spring-boot we can enable it natively in Tomcat or use Jetty GzipFilter in Jetty and Undertow. However Jetty authors themselves do not like filter-based approach to the extent that they have even removed this filter from Jetty-servlet (they left it as a deprecated no-operation filter in Jetty 9.3). 

https://github.com/eclipse/jetty.project/blob/master/jetty-servlets/src/main/java/org/eclipse/jetty/servlets/AsyncGzipFilter.java
https://github.com/eclipse/jetty.project/blob/master/jetty-servlets/src/main/java/org/eclipse/jetty/servlets/GzipFilter.java

Both Jetty and Undertow currently have their native ways to enable gzip compression of responses that can be configured in a similar way. This PR introduces this common configuration properties server.compression.enabled, server.compression.mimeTypes and server.compression.minSize.

At first I wanted to propose this as an alternative way of enabling compression in any of the three containers, but since filter-based way is deprecated in the latest Jetty and will not work after upgrade (Jetty 9.3 requires Java 8, so I assume it will be reasonable to upgrade to it together with Spring 5), maybe this way should be promoted already.

My CLA number is 69320131202043142

cemo commented:
:+1

wilkinsona commented:
Thanks, @isopov. I'd prefer not to have multiple ways to configure the same thing so I think we should explore this approach replacing the existing gzip support. It looks like https://bugs.eclipse.org/bugs/show_bug.cgi?id=470646 is hindering, and perhaps blocking, that at the moment.

You mentioned some API differences between Jetty 9.2 and Jetty 9.3. Ideally, we'd like Boot 1.3 to use Jetty 9.2 by default but also be compatible with 8 (for those stuck with Java 6) and 9.3. It looks like Jetty 8 has `GzipHandler`. If you have it it to hand, can you share some info about the API differences between 9.2 and 9.3 please?

isopov commented:
>>I think we should explore this approach replacing the existing gzip support
Should this replacement be executed at once, or previous options can stay deprecated for some time?

>>can you share some info about the API differences between 9.2 and 9.3 please?
They are straightforward. In 9.3 this class is in the jetty-server artefact with corresponding package - https://github.com/eclipse/jetty.project/blob/jetty-9.3.0.v20150612/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipHandler.java
and in 9.2 this class is in the jetty-servlets artefact - https://github.com/eclipse/jetty.project/blob/jetty-9.2.11.v20150529/jetty-servlets/src/main/java/org/eclipse/jetty/servlets/gzip/GzipHandler.java

So the jetty-servlets dependency that I introduced for the spring-boot in this PR is needed only for Jetty 9.2.

Also the setMimeTypes method in 9.2 is renamed to setIncludedMimeTypes in 9.3 (probably the behaviour is slightly different too, but I haven't noticed that).

>>It looks like Jetty 8 has GzipHandler
Yes. API looks like in the 9.2 branch.
https://github.com/eclipse/jetty.project/blob/jetty-8/jetty-server/src/main/java/org/eclipse/jetty/server/handler/GzipHandler.java

So,  GzipHandler is present in Jetty 8 in the core artefact, moved to jetty-servlets in Jetty 9.2 and moved back (with the third different package name in Jetty 9.3). Should I try to extend the support for all 3? Separate modules for testing than? Current PR support only Jetty 9.2 

Another problem that bothers me is the difference in the featuresets. Actually Jettys featureset is a superset of Tomcat. Undertow has a lower-level API and it seems that all Jettys features can be replicated for Undertow, and maybe even more. Should the spring-boot common compression properties allow to configure only the subset available to Tomcat, or maybe other properties too, but logging (warn, error?) or throwing exception if not-supported features are enabled?

isopov commented:
I have pushed support for all Jetty versions in the second commit for easy review - going to squash it into the main commit later. Not tested it yet with Jetty8 and Jetty 9.3 - going to do it tomorrow.

isopov commented:
I fixed a missprint in this reflection code and tested it - it works with Jetty 8, 9.2 and 9.3 Also minGzipSize works on all three - the problem with it not working in UT is in that content length for documentRoot file resources is not available for GzipHandler in Jettty 9.2. So this code for Jetty is working - it only needs some polish and tests - should it be two separate modules for non-default jetty versions or maybe this tests can be added to existing examples of non-default Jetty versions usage?

Another problem - I have not managed to make gzip working with Tomcat. Will investigte it later.

wilkinsona commented:
@isopov This is great stuff. Thank you.

> Should this replacement be executed at once, or previous options can stay deprecated for some time?

Good question. I'm not sure about the best way to go here. I don't like having multiple ways to do things, but equally I don't like giving people breaking changes to deal with. I should get a chance to discuss this with the other Boot committers tomorrow.

> Should the spring-boot common compression properties allow to configure only the subset available to Tomcat

Yes, I think so. I'd like this to be consistent.

> should it be two separate modules for non-default jetty versions or maybe this tests can be added to existing examples of non-default Jetty versions usage?

Adding some gzip configuration to the existing Tomcat 7 and Jetty 8 samples would be a good starting point. We might also want to add a Jetty 9.3 sample to check that works (even though we don't plan to upgrade to 9.3 just yet (see https://github.com/spring-projects/spring-boot/issues/3108)).

wilkinsona commented:
We discussed this today, and we've decided that the best course of action is to remove the existing HTTP compression support in 1.3 and move everyone onto this approach instead. We'll document how to migrate in the release notes.

wilkinsona commented:
Merged. Thanks for the PR.
